### ISSUE TITLE
XMonad DynamicBars: Pass in dynamicLogString into multiPP. 

### ISSUE THREAD

**Issue by seanstrom** [xmonad/xmonad-contrib#43]
Hi I'm looking for some help on how to implement some functionality with DynamicBars and TopicSpaces. I've been basing my config off of aavogt's archived config and have recently decided to use DynamicBars instead of the initial hand-rolled approach, but there's one small roadblock.

I can't figure out how to combine the output of `TopicSpace.pprWindowSet` with the `DynamicBars.multiPP` function. The reason I want to combine the two is because I would like to show my topic workspaces in order I last visited them. This functionality is achieved in aavogt's config, since the config merges the outputs of `TopicSpace.pprWindowSet` and `dynamicLogString` with a custom function, and that function gets passed into what is the equivalent of `multiPP'`.

At the moment DynamicBars doesn't expose a way for me to pass in my own custom function and I was wondering if another function with the following signature could be exposed:

```haskell
multiPP'' (PP -> X String) -> PP -> PP -> X ()
```

or perhaps someone can tell me of another solution if there's a better way to achieve my goal. In either case I'm happy to get any feedback, thanks for your time.

cc @mathstuf

**Comment by mathstuf** [comment]
Yeah, that seems sensible to me. Since `multiPP'` isn't exposed, you can rename it and not have a weird skip of a single prime for the function.

**Comment by aavogt** [comment]
If you leave off the type signature, multiPP' has a more general type:

```haskell
multiPP' :: (a -> X String) -> a -> a -> [Handle] -> X ()

-- so maybe it should have a name that doesn't mention PP,
-- and maybe a Bool is easier to deal with

-- | @logScreens log@ runs the supplied action (@log@) on each screen,
-- resetting the XState before running each action. The
-- focused screen's update to the XState is retained
logScreens :: (Bool -- ^ is the screen being logged also focused?
                -> ScreenId -> X ())
                -> X ()
logScreens = -- almost what multiPP' is
```

**Comment by seanstrom** [comment]
@aavogt what are some of the advantages of receiving a `ScreenId` in the function? Is that used so we can output to the correct status bar?

**Comment by mathstuf** [comment]
ScreenId is good to have. It's available and could be useful (window list on one screen, info on the other).

**Comment by aavogt** [comment]
@seanstrom yes it would be convenient, but I guess that argument could come from `gets (W.screen . W.current . windowset) :: X ScreenId` instead.

**Comment by mathstuf** [comment]
That gets the currently focused window's ScreenId, not the bar's screen though.

**Comment by aavogt** [comment]
The `put st{ windowset = W.view ws $ windowset st }` line in multiPP' should be changing the ScreenId you get from the `gets (W.screen . W.current . windowset)`

**Comment by seanstrom** [comment]
I started work on this and I added a function for my use case named `multiPPFormat`. I'm not sure if that's the best name, but I thought it did okay at expressing the intent.

I've also started taking a look into how to rewrite the `multiPP` functions into something more along the lines of `logScreens`, but that may take me longer since I'm still trying to understand the rest of the code. Let me know if we're interested in something like `multiPPFormat` and I can prepare a patch. Otherwise I'll keep working on the refactor.

Side note:
I've been spending the better part of my day getting a development environment setup, any tips on that? And how do we normally go about testing the full functionality after compilation?
